[Frontend] -frontend -emit-loaded-module-trace.

huonw opened the pull request:
The -frontend jobs can output a JSON file summarizing the
swiftmodules (etc.) they load during compilation.

huonw committed:
[Frontend] -frontend -emit-loaded-module-trace.

The -frontend jobs can output a JSON file summarizing the
swiftmodules (etc.) they load during compilation.

jrose-apple commented on lib/Frontend/LoadedModuleTrace.cpp:
```diff
@@ -0,0 +1,47 @@
+
+namespace swift {
+namespace json {
+template <typename T> struct ArrayTraits<std::vector<T>> {
```
This seems like it should just be sunk into JSONSerialization. Otherwise someone will end up adding their own.

huonw replied:
Indeed. Someone like me.

jrose-apple replied:
We own JSONSerialization for now, go ahead and push it down.

huonw committed:
[Driver] Main compiler tells frontend to -emit-loaded-module-trace.

jrose-apple commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -132,6 +142,67 @@ static bool emitMakeDependencies(DiagnosticEngine &diags,
+    if (ext == SERIALIZED_MODULE_EXTENSION) {
+      swiftModules.push_back(realPath);
+    } else {
+      assert(ext == clang::api_notes::BINARY_APINOTES_EXTENSION);
```
This is a bizarre thing to put in the trace file. apinotesc files are mostly generated on demand.

What about things in the list of dependencies that are neither modules nor apinotesc files?

huonw replied:
I'm mostly cargoculting here: the only things that turned up in the list of dependencies for the things I tested were modules and apinotesc's. It sounds like you expect we only need the former out of those two; and that there's other possibilities...

jrose-apple replied:
Headers can also appear here, if you import any non-system headers.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -199,6 +199,16 @@ def emit_dependencies : Flag<["-"], "emit-dependencies">,
   HelpText<"Emit basic Make-compatible dependencies files">;
 
+def emit_loaded_module_trace : Flag<["-"], "emit-loaded-module-trace">,
+  Flags<[FrontendOption, NoInteractiveOption, DoesNotAffectIncrementalBuild]>,
```
This probably shouldn't be DoesNotAffectIncrementalBuild, since adding the option means you need to rebuild.

huonw committed:
[Driver] Support the SWIFT_LOADED_MODULE_TRACE_PATH env var.

This overrides -emit-loaded-module-trace (etc.).

huonw committed:
[test] C header, multifile, is-one-line and transitive dependency tests for -emit-loaded-module-trace.

huonw committed:
[Frontend] Buffer loaded module trace in memory and append "atomically".

huonw commented:
Updated:

- addressed the review,
- included the logic for getting exactly one frontend job to emit this file,
- added B&I's ask for an environment variable to direct the file (easier than threading arguments around)
- switched file generation to be appending, again at B&I's request
- more tests

huonw commented:
@swift-ci Please smoke test

jrose-apple commented:
Ah, I still have a preference for "threading arguments around". Frontend jobs are supposed to depend as little as possible on the environment, so that the backtrace you get from a crash is as reproducible as possible.

jrose-apple commented:
Oh, never mind, I see you did that already. :-)

huonw commented:
Ah, yeah, I was confusing: internally arguments are threaded around, but higher level systems that want to use this may be quite removed from the top-level compiler invocations so the latter will check the env var.

huonw commented:
@swift-ci Please smoke test
